File: ocaml/AC17.ml
let compare_pair cmp_a cmp_b (a1, b1) (a2, b2) =
  match cmp_a a1 a2 with
  | 0 -> cmp_b b1 b2
  | cmp -> cmp

module Direction = struct
  type t = Horizontal | Vertical

  let compare a b =
    match a, b with
    | Horizontal, Horizontal -> 0
    | Horizontal, Vertical -> -1
    | Vertical, Horizontal -> 1
    | Vertical, Vertical -> 0

  let turn direction =
    match direction with
    | Horizontal -> Vertical
    | Vertical -> Horizontal
end

module Coords = struct
  type t = {
      x: int;
      y: int;
    }

  let compare a b =
    compare_pair Int.compare Int.compare (a.x, a.y) (b.x, b.y)

  let size matrix =
    { x = Array.length matrix.(0); y = Array.length matrix }

  let ( + ) a b =
    { x = a.x + b.x; y = a.y + b.y }

  let in_bounds { x = width; y = height } { x; y } =
    x >= 0 && y >= 0 && x < width && y < height

  let get matrix { x; y } =
    matrix.(y).(x)

  let of_direction (direction : Direction.t) =
    match direction with
    | Horizontal -> [{ x = -1; y = 0 }; { x = 1; y = 0 }]
    | Vertical -> [{ x = 0; y = -1 }; { x = 0; y = 1 }]
end

let parse_line line =
  Array.of_seq (Seq.map
    (fun c -> Option.get (Common.int_of_digit c)) (String.to_seq line))

module State = struct
  type t = {
      pos: Coords.t;
      direction: Direction.t;
    }

  let compare a b =
    compare_pair Coords.compare Direction.compare
      (a.pos, a.direction) (b.pos, b.direction)
end

module HeatlossState = struct
  type t = {
      heatloss: int;
      state: State.t;
    }

  let compare a b =
    compare_pair Int.compare State.compare
      (a.heatloss, a.state) (b.heatloss, b.state)
end

module StateMap = Map.Make (State)

module HeatlossStateMap = Map.Make (HeatlossState)

let rec find_path_loop min_turn max_forward grid tgt heap_map heatloss_map =
  let heatloss_state, () = HeatlossStateMap.min_binding heap_map in
  let heap_map = HeatlossStateMap.remove heatloss_state heap_map in
  if heatloss_state.state.pos = tgt then
    heatloss_state.heatloss
  else
    begin
      let heatloss = StateMap.find heatloss_state.state heatloss_map in
      if heatloss <> heatloss_state.heatloss then
        find_path_loop min_turn max_forward grid tgt heap_map heatloss_map
      else
        begin
          let direction = Direction.turn heatloss_state.state.direction in
          let offsets = Coords.of_direction direction in
          let heap_map, heatloss_map =
            List.fold_left (fun (heap_map, heatloss_map) offset ->
              let rec make_step step pos heatloss heap_map heatloss_map =
                if step > max_forward then
                  heap_map, heatloss_map
                else
                  begin
                    let pos = Coords.(pos + offset) in
                    if Coords.in_bounds (Coords.size grid) pos then
                      begin
                        let heatloss = heatloss + Coords.get grid pos in
                        let heap_map, heatloss_map =
                          if step >= min_turn then
                            begin
                              let state = State.{ pos; direction } in
                              match StateMap.find state heatloss_map with
                              | exception Not_found ->
                                 HeatlossStateMap.add { heatloss; state } ()
                                   heap_map,
                                 StateMap.add state heatloss heatloss_map
                              | heatloss' ->
                                 if heatloss' <= heatloss then
                                   heap_map, heatloss_map
                                 else
                                   HeatlossStateMap.add { heatloss; state } ()
                                     heap_map,
                                   StateMap.add state heatloss heatloss_map
                            end
                          else
                            heap_map, heatloss_map in
                        make_step (step + 1) pos heatloss heap_map heatloss_map
                      end
                    else
                      heap_map, heatloss_map
                  end in
              make_step 1 heatloss_state.state.pos heatloss heap_map
                heatloss_map)
              (heap_map, heatloss_map) offsets in
          find_path_loop min_turn max_forward grid tgt heap_map heatloss_map
        end
    end

let find_path ~min_turn ~max_forward grid from tgt =
  let heap_map = HeatlossStateMap.of_list [
    { heatloss = 0; state = { pos = from; direction = Horizontal }}, ();
    { heatloss = 0; state = { pos = from; direction = Vertical }}, ();
  ] in
  let heatloss_map = StateMap.of_list [
    { pos = from; direction = Horizontal }, 0;
    { pos = from; direction = Vertical }, 0;
  ] in
  find_path_loop min_turn max_forward grid tgt heap_map heatloss_map

let () =
  let grid = Array.of_seq (Seq.map parse_line (Common.input_lines ())) in
  let result_part1 =
    find_path grid { x = 0; y = 0 } Coords.(size grid + { x = -1; y = -1 })
      ~min_turn:0 ~max_forward:3 in
  Format.eprintf "Part 1: %d@." result_part1;
  let result_part2 =
    find_path grid { x = 0; y = 0 } Coords.(size grid + { x = -1; y = -1 })
      ~min_turn:4 ~max_forward:10 in
  Format.eprintf "Part 2: %d@." result_part2

File: rust/src/bin/ac17.rs
use advent_of_code::{
    Error, Result, Matrix2D, Coords2D, Zero, matrix_from_lines
};

#[derive(Clone, Copy,PartialEq, Eq, PartialOrd, Ord, Hash)]
enum Direction {
    Horizontal,
    Vertical,
}

impl Direction {
    fn turn(self) -> Self {
        match self {
            Direction::Horizontal => Direction::Vertical,
            Direction::Vertical => Direction::Horizontal,
        }
    }
}

#[derive(Clone, PartialEq, Eq)] 
struct State {
    heat_loss: u64,
    pos: Coords2D<usize>,
    dir: Direction,
}

impl State {
    fn as_tuple(&self) -> (std::cmp::Reverse<u64>, usize, usize, Direction) {
        (std::cmp::Reverse(self.heat_loss), self.pos.x, self.pos.y, self.dir)
    }
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.as_tuple().cmp(&other.as_tuple())
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

fn find_path(
    grid: &Matrix2D<u32>, min_turn: u64, max_forward: u64,
    from: Coords2D<usize>, to: Coords2D<usize>
) -> Result<u64> {
    let mut heap = std::collections::BinaryHeap::new();
    let mut heat_losses = std::collections::HashMap::new();
    heap.push(State { heat_loss: 0, pos: from, dir: Direction::Horizontal });
    heat_losses.insert((from, Direction::Horizontal), 0);
    heap.push(State { heat_loss: 0, pos: from, dir: Direction::Vertical });
    heat_losses.insert((from, Direction::Vertical), 0);
    loop {
        let state = heap.pop().ok_or("Empty queue")?;
        if state.pos == to {
            return Ok(state.heat_loss);
        }
        let heat_loss =
            heat_losses.get(&(state.pos, state.dir))
            .ok_or("Position not found")?;
        if *heat_loss != state.heat_loss {
            continue;
        }
        let dir = state.dir.turn();
        let offsets: [Coords2D<isize>; 2] =
            match dir {
                Direction::Vertical => [Coords2D::UP, Coords2D::DOWN],
                Direction::Horizontal => [Coords2D::LEFT, Coords2D::RIGHT],
            };
        for offset in offsets {
            let mut pos = state.pos;
            let mut heat_loss = state.heat_loss;
            for step in 1..=max_forward {
                let Some(next_pos) = pos.advance(grid.into(), offset)
                else { break };
                pos = next_pos;
                heat_loss += *pos.get(grid) as u64;
                if step >= min_turn {
                    match heat_losses.entry((pos, dir)) {
                        std::collections::hash_map::Entry::Vacant(vacant) => {
                            vacant.insert(heat_loss);
                        }
                        std::collections::hash_map::Entry::Occupied(mut occupied) => {
                            let cell = occupied.get_mut();
                            if *cell <= heat_loss {
                                continue;
                            }
                            *cell = heat_loss;
                        }
                    }
                    heap.push(State { heat_loss, pos, dir });
                }
            }
        }
    }
}

fn main() -> Result<()> {
    let char_grid = matrix_from_lines(std::io::stdin().lines())?;
    let grid: Matrix2D<u32> = char_grid.iter().map(|line|
      line.iter().map(|c| c.to_digit(10)
          .ok_or(Error::from(format!("Invalid digit {c}"))))
      .collect::<Result<_>>()
    ).collect::<Result<_>>()?;
    let from: Coords2D<usize> = Coords2D::ZERO;
    let to: Coords2D<usize> =
        Coords2D::from(&grid).checked_add_signed(Coords2D::UP_LEFT)
        .ok_or("Empty grid")?;
    let result_part1 = find_path(&grid, 0, 3, from, to)?;
    println!("Part 1: {result_part1}");
    let result_part2 = find_path(&grid, 4, 10, from, to)?;
    println!("Part 2: {result_part2}");
    Ok(())
}
